deep_speech_2/ctc_beam_search_decoder.py

```diff
@@ -0,0 +1,192 @@
+## This is a prototype of ctc beam search decoder
+
+import copy
+import random
+import numpy as np
+
+# vocab = blank + space + English characters
+#vocab = ['-', ' '] + [chr(i) for i in range(97, 123)]
+
+vocab = ['-', '_', 'a']
+
+
+def ids_list2str(ids_list):
+    ids_str = [str(elem) for elem in ids_list]
+    ids_str = ' '.join(ids_str)
+    return ids_str
+
+
+def ids_id2token(ids_list):
+    ids_str = ''
+    for ids in ids_list:
+        ids_str += vocab[ids]
+    return ids_str
+
+
+def language_model(ids_list, vocabulary):
+    # lookup ptb vocabulary
+    ptb_vocab_path = "./data/ptb_vocab.txt"
+    sentence = ''.join([vocabulary[ids] for ids in ids_list])
+    words = sentence.split(' ')
+    last_word = words[-1]
+    with open(ptb_vocab_path, 'r') as ptb_vocab:
+        f = ptb_vocab.readline()
+        while f:
+            if f == last_word:
+                return 1.0
+            f = ptb_vocab.readline()
+    return 0.0
+
+
+def ctc_beam_search_decoder(input_probs_matrix,
+                            beam_size,
+                            vocabulary,
+                            max_time_steps=None,
+                            lang_model=language_model,
+                            alpha=1.0,
+                            beta=1.0,
+                            blank_id=0,
+                            space_id=1,
+                            num_results_per_sample=None):
+    '''
+    Beam search decoder for CTC-trained network,  adapted from Algorithm 1 
+    in https://arxiv.org/abs/1408.2873.
+
+    :param input_probs_matrix: probs matrix for input sequence, row major
+    :type input_probs_matrix: 2D matrix. 
+    :param beam_size: width for beam search
```

Comment:
1. width --> Width
2. Don't forget put a dot in the end of a sentence. Same below.